File: app/service/auth_proxy.rb
# app/service/auth_proxy.rb
require 'httparty'

class AuthProxy
  include HTTParty

  base_uri ENV['AUTH_SERVICE_URL'] || 'http://localhost:3001'

  # ✅ Login - obtener access y refresh tokens
  def self.login(username:, password:)
    response = post(
      '/auth/login',
      body: { username: username, password: password }.to_json,
      headers: { 'Content-Type' => 'application/json' }
    )

    if response.success?
      {
        success: true,
        access_token: response['access'] || response['access_token'],
        refresh_token: response['refresh'] || response['refresh_token'],
        user: response['user']
      }
    else
      {
        success: false,
        error: response['message'] || 'Login falló'
      }
    end
  rescue StandardError => e
    { success: false, error: "Error conectando al Auth Service: #{e.message}" }
  end

  # ✅ Registro - crear nuevo usuario
  def self.register(username:, email:, password:, role: 'user')
    response = post(
      '/auth/register',
      body: {
        username: username,
        email: email,
        password: password,
        role: role
      }.to_json,
      headers: { 'Content-Type' => 'application/json' }
    )

    if response.success?
      {
        success: true,
        user: response['user']
      }
    else
      {
        success: false,
        error: response['message'] || 'Registro falló'
      }
    end
  rescue StandardError => e
    { success: false, error: "Error conectando al Auth Service: #{e.message}" }
  end

  # ✅ Logout - revocar tokens
  def self.logout(access_token:, refresh_token:)
    response = post(
      '/auth/logout',
      body: { refresh: refresh_token }.to_json,
      headers: {
        'Content-Type' => 'application/json',
        'Authorization' => "Bearer #{access_token}"
      }
    )

    if response.success?
      { success: true, message: 'Logout exitoso' }
    else
      { success: false, error: response['message'] || 'Logout falló' }
    end
  rescue StandardError => e
    { success: false, error: "Error conectando al Auth Service: #{e.message}" }
  end

  # ✅ Refresh - obtener nuevo access token
  def self.refresh(refresh_token:)
    response = post(
      '/auth/refresh',
      body: { refresh: refresh_token }.to_json,
      headers: { 'Content-Type' => 'application/json' }
    )

    if response.success?
      {
        success: true,
        access_token: response['access'] || response['access_token'],
        refresh_token: response['refresh'] || refresh_token
      }
    else
      { success: false, error: response['message'] || 'Refresh falló' }
    end
  rescue StandardError => e
    { success: false, error: "Error conectando al Auth Service: #{e.message}" }
  end

  # ✅ Validate - verificar token contra el Auth Service (para revocación)
  def self.validate(access_token:)
    response = post(
      '/auth/validate',
      body: { token: access_token }.to_json,
      headers: { 'Content-Type' => 'application/json' }
    )

    if response.success?
      {
        success: true,
        data: response.parsed_response
      }
    else
      {
        success: false,
        error: response.parsed_response && response.parsed_response['message'] || 'Token inválido'
      }
    end
  rescue StandardError => e
    { success: false, error: "Error validando token: #{e.message}" }
  end
end

File: SalumedX/login/views/auth_proxy.py
"""
Vista proxy para redirigir login/registro al auth-service
El auth-service (puerto 8001) devuelve tokens JWT que luego son validados localmente
"""
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import AllowAny
from rest_framework.response import Response
import requests
from django.conf import settings

AUTH_SERVICE_URL = getattr(settings, 'AUTH_SERVICE_URL', 'http://localhost:8001')


@api_view(['POST'])
@permission_classes([AllowAny])
def signin_proxy(request):
    """
    Proxy para login - redirige al auth-service
    
    El auth-service devuelve:
    {
        "access": "<JWT_TOKEN>",
        "refresh": "<REFRESH_TOKEN>",
        "user": {...}
    }
    
    Este token debe ser usado en requests posteriores:
    Authorization: Bearer <JWT_TOKEN>
    """
    try:
        response = requests.post(
            f'{AUTH_SERVICE_URL}/auth/login',
            json=request.data,
            timeout=5
        )
        
        data = response.json()
        
        # Agregar información sobre el uso del token
        if response.status_code == 200 and 'access' in data:
            data['token_info'] = {
                'usage': 'Incluir en header: Authorization: Bearer <access_token>',
                'expires_in': '5 horas',
                'refresh_usage': 'Usar refresh token en /token/refresh/ para obtener nuevo access token'
            }
        
        return Response(data, status=response.status_code)
        
    except requests.exceptions.Timeout:
        return Response({
            'error': 'Auth service timeout',
            'detail': 'El servicio de autenticación no respondió a tiempo'
        }, status=504)
    except requests.exceptions.ConnectionError:
        return Response({
            'error': 'Auth service no disponible',
            'detail': 'No se pudo conectar al servicio de autenticación'
        }, status=503)
    except requests.exceptions.RequestException as e:
        return Response({
            'error': 'Auth service error',
            'detail': str(e)
        }, status=503)


@api_view(['POST'])
@permission_classes([AllowAny])
def signup_proxy(request):
    """
    Proxy para registro - redirige al auth-service
    
    El auth-service devuelve:
    {
        "access": "<JWT_TOKEN>",
        "refresh": "<REFRESH_TOKEN>",
        "user": {...}
    }
    
    Después del registro, el usuario recibe tokens automáticamente
    """
    try:
        response = requests.post(
            f'{AUTH_SERVICE_URL}/auth/register',
            json=request.data,
            timeout=5
        )
        
        data = response.json()
        
        # Agregar información sobre el uso del token
        if response.status_code == 201 and 'access' in data:
            data['token_info'] = {
                'usage': 'Incluir en header: Authorization: Bearer <access_token>',
                'expires_in': '5 horas',
                'refresh_usage': 'Usar refresh token en /token/refresh/ para obtener nuevo access token'
            }
        
        return Response(data, status=response.status_code)
        
    except requests.exceptions.Timeout:
        return Response({
            'error': 'Auth service timeout',
            'detail': 'El servicio de autenticación no respondió a tiempo'
        }, status=504)
    except requests.exceptions.ConnectionError:
        return Response({
            'error': 'Auth service no disponible',
            'detail': 'No se pudo conectar al servicio de autenticación'
        }, status=503)
    except requests.exceptions.RequestException as e:
        return Response({
            'error': 'Auth service error',
            'detail': str(e)
        }, status=503)
